browser/mapping.go

```diff
@@ -648,7 +648,22 @@ func mapBrowserContext(vu moduleVU, bc api.BrowserContext) mapping {
 		"setOffline":         bc.SetOffline,
 		"storageState":       bc.StorageState,
 		"unroute":            bc.Unroute,
-		"waitForEvent":       bc.WaitForEvent,
+		"waitForEvent": func(event string, optsOrPredicate goja.Value) *goja.Promise {
+			ctx := vu.Context()
+			return k6ext.Promise(ctx, func() (result any, reason error) {
+				resp, err := bc.WaitForEvent(event, optsOrPredicate)
+				if err != nil {
+					return nil, err //nolint:wrapcheck
+				}
+
+				p, ok := resp.(api.Page)
+				if !ok {
+					panicIfFatalError(ctx, fmt.Errorf("response object is not a page: %w", k6error.ErrFatal))
+				}
```

Comment:
What's the difference between returning an error and panicking here? We discussed this before, but please remind me. Thanks!

Reply:
The reason i didn't want to use `k6ext.Abort` from here is to use `panicIfFatalError` which could have other cleanup functionality in the future. At the moment `panicIfFatalError` doesn't do anything but call `k6ext.Abort`.

In general it tries to help isolate where an `k6ext.Abort` can be called from, and so it encourages us to bubble up errors from downstream back up. If we wrap an error as `FatelError` then it will abort at the mapping layer instead of us having to call abort/panic in the business logic area of the codebase.

In this case, maybe it's ok to call `k6ext.Abort` from here instead of first wrapping it in `FatelError` and then calling `panicIfFatalError`.

The reason we wanted something like this in the first place was due to https://github.com/grafana/xk6-browser/issues/688

Let me know how you feel about this.

Reply:
Ah, right, so the reason I prefer an abort here over a plain error is because if we get something other than `api.page` then something has gone very wrong in the implementation, and requires us to look into. Happy change it to a error though.

Reply:
> Happy to change it to an error though.

Yep, please 👍

I understand your reasoning better now, thanks. Returning something other than `api.Page` deserves a panic in the business logic code, which will be a programmer error (that's what panics are for—truly exceptional situations). For usual error handling, I think returning an error is fine here.

Reply:
I have mixed thoughts here (see my [comment](https://github.com/grafana/xk6-browser/pull/1042#discussion_r1334016282) kind of related to this).
I guess the first thing that we should decide is if receiving an event data that does not comply with a `Page` should produce a panic or not.

Many times we used panic because there was not a "direct channel" to communicate the error to the user, but in this case we do have it as we are in the context of a promise, so would it make sense to return every error, even if it is an unexpected one, through this promise?

Reply:
I believe unexpected programming mistakes should cause a panic (i.e., [1](https://cs.opensource.google/go/go/+/refs/tags/go1.21.1:src/crypto/tls/conn.go;l=252-274), [2](https://github.com/golang/go/blob/master/src/cmd/compile/internal/base/flag.go#L357-L433)), not an error. Errors are for expected and usual working conditions, like a file is not found, or a network is down. That's how I see it, and I'm happy to leave the final decision to you 👍

Reply:
Yeah, i agree with both of what you are saying. Here's an update:

I somehow missed this type cast check that @ka3de pointed out [here](https://github.com/grafana/xk6-browser/pull/1042#discussion_r1334010235) 🤦 So that has now been resolved.

At the moment, `resp` could either be a nil `page` or a valid `page`. I feel that it's probably still valid to panic here still though even though we're now correctly checking the casting downstream from here. WDYT?

Reply:
I'm leaning towards the idea that type mistakes likely come from programming errors, and it might be best to _use the built-in `panic`_. However, I'm open to other approaches and happy to go with what the team thinks is best!

Reply:
You mean panic in the `BrowserContext.waitForEvent` implementation? The reason not to do that would be to avoid panicking on goroutine other than the "main" goroutine.

If you're wondering why use `panicIfFatalError` instead of just the `k6ext.Abort` method directly, i'd be happy to do that too here too if that's the consensus 🙂